### PR TITLE
ft: enable versioning to aws backend for mpus

### PR DESCRIPTION
**Pull request by electrachong**

### CODE REVIEW

File: lib/api/completeMultipartUpload.js [hunk 1]
```diff
@@ -18,8 +18,7 @@ const { versioningPreprocessing, checkQueryVersionId }
 const metadata = require('../metadata/wrapper');
 const services = require('../services');
 const { metadataValidateBucketAndObj } = require('../metadata/metadataUtils');
-const { checkAzureBackendMatch, skipMpuPartProcessing } =
```
Comment (electrachong) [inline lib/api/completeMultipartUpload.js]: maybe need to remove checkAzureBackendMatch if it's not used anywhere else

File: lib/data/wrapper.js [hunk 2]
```diff
@@ -59,6 +59,26 @@ if (config.backends.data === 'mem') {
...
+    const skipMethods = { PUT: true, POST: true };
+    if (!Array.isArray(locations) || !locations[0] ||
+        !locations[0].dataStoreType) {
+        return false;
+    }
+    const isSkipBackend = skipBackend[locations[0].dataStoreType];
+    const isMatchingBackends =
+        locations[0].dataStoreName === newObjDataStoreName;
+    const isSkipMethod = skipMethods[requestMethod];
+    if (!isSkipBackend || !isMatchingBackends || !isSkipMethod) {
```
Comment (bennettbuchanan) [inline lib/data/wrapper.js]: Minor: the result can just be returned instead of using an if condition.

File: lib/data/external/AwsClient.js [hunk 3]
```diff
@@ -327,6 +329,12 @@ class AwsClient {
                   `AWS: ${err.message}`)
                 );
             }
+            if (!completeMpuRes.VersionId) {
```
Comment (bennettbuchanan) [inline lib/data/external/AwsClient.js]: Does versioning have to be enabled on the bucket to use multiple backend routes? If so, I guess I would not expect the complete MPU operation to succeed if also receiving this error. But maybe we want to be flexible here.

Comment (electrachong) [reply lib/data/external/AwsClient.js]: Versioning has to be enabled on AWS to use multiple backends with AWS. That's why we return an error if no version id is returned (means versioning was suspended at some point)

File: lib/data/wrapper.js (hunk 2 above)
Comment (electrachong) [reply lib/data/wrapper.js]: Thanks!

File: lib/data/external/AwsClient.js (hunk 3 above)
Comment (LaurenSpiegel) [reply lib/data/external/AwsClient.js]: are we doing the same on puts?

File: lib/data/wrapper.js [hunk 4]
```diff
@@ -59,6 +59,23 @@ if (config.backends.data === 'mem') {
  */
 const MAX_RETRY = 2;
 
+// This check is done because on a put, complete mpu or copy request to AWS,
+// if the object already exists on AWS, the existing object should not be
```
Comment (LaurenSpiegel) [inline lib/data/wrapper.js]: what about azure?

File: lib/data/wrapper.js [hunk 5]
```diff
@@ -59,6 +59,23 @@ if (config.backends.data === 'mem') {
...
+// deleted, which is the functionality for all other backends
+function _shouldSkipDelete(locations, requestMethod, newObjDataStoreName) {
+    const skipBackend = externalBackends;
```
Comment (LaurenSpiegel) [inline lib/data/wrapper.js]: what does this do?

File: lib/data/wrapper.js [hunk 6]
```diff
@@ -59,6 +59,23 @@ if (config.backends.data === 'mem') {
...
+    const skipMethods = { PUT: true, POST: true };
+    if (!Array.isArray(locations) || !locations[0] ||
+        !locations[0].dataStoreType) {
+        return false;
```
Comment (LaurenSpiegel) [inline lib/data/wrapper.js]: why do you have to still delete if there is no data?

File: lib/data/external/AwsClient.js (hunk 3 above)
Comment (electrachong) [reply lib/data/external/AwsClient.js]: @LaurenSpiegel yes, we are

File: lib/data/wrapper.js (hunk 4 above)
Comment (electrachong) [reply lib/data/wrapper.js]: Added Azure to the comment.

File: lib/data/wrapper.js (hunk 5 above)
Comment (electrachong) [reply lib/data/wrapper.js]: Nothing much, it's leftover from the original check. I removed it

File: lib/data/wrapper.js (hunk 6 above)
Comment (electrachong) [reply lib/data/wrapper.js]: The batch delete uses an async.eachLimit method so it won't do anything for an empty array even if we return `false`.
I return early if it's an empty array here so we won't get an error below accessing the `dataStoreType` and `dataStoreName` properties of the 0th index.

**Comment by electrachong** [comment]
@ironman-machine try

**Comment by electrachong** [comment]
end to end passed here: http://ci.ironmann.io/gh/scality/Integration/15988#build-parameters

**Pull request merged**